False positive with JSpecify nullable array in record with a custom constructor

Hi, I'm having a strange situation where NullAway incorrectly reports that a `@NonNull` value is required. I've managed to reduce it to the following test class : it fails on the static method, but only if the public custom constructor is defined (if I remove the custom constructor, NullAway reports no problem with the static method)

```java
import java.util.Objects;
import org.jspecify.annotations.Nullable;

public record Test(Object type, String filename, byte @Nullable [] value) {

  public static Test file(final String filename) {
    return new Test(new Object(), filename, null);
  }

  public Test {
    Objects.requireNonNull(type, "type");
    Objects.requireNonNull(filename, "filename");
  }
}

```

```
FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':....:compileJava' (registered by plugin class 'org.gradle.api.plugins.JavaBasePlugin').
> Compilation failed; see the compiler output below.
  /.../Test.java:9: error: [NullAway] passing @Nullable parameter 'null' where @NonNull is required
      return new Test(new Object(), filename, null);
                                              ^
      (see http://t.uber.com/nullaway )
  1 error
```

I'm using NullAway 0.13.7, with errorprone 2.50.0

--- a/nullaway/src/main/java/com/uber/nullaway/NullabilityUtil.java
+++ b/nullaway/src/main/java/com/uber/nullaway/NullabilityUtil.java
@@ -356,14 +356,20 @@ public class NullabilityUtil {
   public static Stream<? extends AnnotationMirror> getAllAnnotationsForParameter(
       Symbol.MethodSymbol symbol, int paramInd, Config config) {
     Symbol.VarSymbol varSymbol = symbol.getParameters().get(paramInd);
-    return Stream.concat(
-        varSymbol.getAnnotationMirrors().stream(),
-        symbol.getRawTypeAttributes().stream()
-            .filter(
-                t ->
-                    t.position.type.equals(TargetType.METHOD_FORMAL_PARAMETER)
-                        && t.position.parameter_index == paramInd
-                        && NullabilityUtil.isDirectTypeUseAnnotation(t, symbol, config)));
+    // On modern javac versions, type-use annotations are attached directly to the parameter's
+    // type. Keep reading raw attributes as well for backward compatibility with older javac
+    // behavior and legacy annotation-location handling.
+    Stream<? extends AnnotationMirror> typeUseAnnotations =
+        Stream.concat(
+                varSymbol.type.getAnnotationMirrors().stream(),
+                symbol.getRawTypeAttributes().stream()
+                    .filter(
+                        t ->
+                            t.position.type.equals(TargetType.METHOD_FORMAL_PARAMETER)
+                                && t.position.parameter_index == paramInd
+                                && NullabilityUtil.isDirectTypeUseAnnotation(t, symbol, config)))
+            .distinct();
+    return Stream.concat(varSymbol.getAnnotationMirrors().stream(), typeUseAnnotations);
   }
 
   /**

--- a/nullaway/src/test/java/com/uber/nullaway/TypeUseAnnotationsTests.java
+++ b/nullaway/src/test/java/com/uber/nullaway/TypeUseAnnotationsTests.java
@@ -412,12 +412,19 @@ public class TypeUseAnnotationsTests extends NullAwayTestsBase {
         .addSourceLines(
             "Test.java",
             """
-            import java.util.Objects;
             import org.jspecify.annotations.*;
             @NullMarked
-            public record Test(byte @Nullable [] value) {
+            public record Test(String name, byte @Nullable [] value, Object marker) {
               public static Test file() {
-                return new Test(null);
+                return new Test("file", null, new Object());
+              }
+              public static Test nullName() {
+                // BUG: Diagnostic contains: passing @Nullable parameter 'null' where @NonNull is required
+                return new Test(null, new byte[0], new Object());
+              }
+              public static Test nullMarker() {
+                // BUG: Diagnostic contains: passing @Nullable parameter 'null' where @NonNull is required
+                return new Test("file", new byte[0], null);
               }
               public Test {
               }

--- a/nullaway/src/test/java/com/uber/nullaway/TypeUseAnnotationsTests.java
+++ b/nullaway/src/test/java/com/uber/nullaway/TypeUseAnnotationsTests.java
@@ -406,6 +406,26 @@ public class TypeUseAnnotationsTests extends NullAwayTestsBase {
         .doTest();
   }
 
+  @Test
+  public void recordConstructorParameter() {
+    defaultCompilationHelper
+        .addSourceLines(
+            "Test.java",
+            """
+            import java.util.Objects;
+            import org.jspecify.annotations.*;
+            @NullMarked
+            public record Test(byte @Nullable [] value) {
+              public static Test file() {
+                return new Test(null);
+              }
+              public Test {
+              }
+            }
+            """)
+        .doTest();
+  }
+
   private CompilationTestHelper makeLegacyModeHelper() {
     return makeTestHelperWithArgs(
         Arrays.asList(

--- a/nullaway/src/test/java/com/uber/nullaway/jspecify/JSpecifyArrayTests.java
+++ b/nullaway/src/test/java/com/uber/nullaway/jspecify/JSpecifyArrayTests.java
@@ -817,6 +817,28 @@ public class JSpecifyArrayTests extends NullAwayTestsBase {
         .doTest();
   }
 
+  @Test
+  public void recordConstructorParameter() {
+    makeHelper()
+        .addSourceLines(
+            "Test.java",
+            """
+            import java.util.Objects;
+            import org.jspecify.annotations.*;
+            @NullMarked
+            public record Test(Object type, String filename, byte @Nullable [] value) {
+              public static Test file(final String filename) {
+                return new Test(new Object(), filename, null);
+              }
+              public Test {
+                Objects.requireNonNull(type, "type");
+                Objects.requireNonNull(filename, "filename");
+              }
+            }
+            """)
+        .doTest();
+  }
+
   private CompilationTestHelper makeHelper() {
     return makeTestHelperWithArgs(
         JSpecifyJavacConfig.withJSpecifyModeArgs(

